pkg/antctl/raw/supportbundle/command.go

```diff
@@ -641,9 +654,157 @@ func processResults(resultMap map[string]error, dir string) error {
 		err = writeFailedNodes(dir, failedNodes)
 	}
 
+	// download logs from kubernetes api
+	if failedNodes != nil || controllerFail {
+		err := downloadPodInfoFromKubernetes(antreaClientset, k8sClient, failedNodes, controllerFail, dir)
+		if err != nil {
+			fmt.Println("Failed to download logs from kubernetes api: " + err.Error())
+		} else {
+			allFailed = false
+		}
+	}
+
 	if allFailed {
 		return fmt.Errorf("no data was collected: %s", genErrorMsg(resultMap))
 	} else {
 		return err
 	}
 }
+
+// downloadLogsFromKubernetes will try to download pod logs from kubernetes api for failed nodes and controller, as well as controllerinfo/agentinfo...
+func downloadPodInfoFromKubernetes(antreaClientset antrea.Interface, k8sClient kubernetes.Interface, failedNodes []string, isControllerFail bool, dir string) error {
```

Comment:
Having codes collecting controller info and agent info in same function doesn't seem save many shared code and causes unnecessary processing when only one of them fails. For example, all agentinfo and agentPods are fetched when there is no failed node.  Perhaps it's cleaner to have two functions:
```
func downloadControlleBundleFromKubernetes(antreaClientset antrea.Interface, k8sClient kubernetes.Interface, dir string) error {
	var errors []error
	controllerInfo, err := antreaClientset.CrdV1beta1().AntreaControllerInfos().Get(context.TODO(), "antrea-controller", metav1.GetOptions{})
	if err != nil {
		...
	} else {
		...
	}
	
	pods, err := k8sClient.CoreV1().Pods("kube-system").List(context.TODO(), metav1.ListOptions{
		ResourceVersion: "0",
		LabelSelector:   "app=antrea,component=antrea-controller",
	})
        ...
	for _, pod := range pods.Items {
		tmpDir, err := afero.TempDir(defaultFS, "", "bundle_tmp_")
		...
		err = downloadPodLogs(k8sClient, "controller", pod.Namespace, pod.Name, []string{"antrea-controller"}, dir, tmpDir)
		...
	}
	return errors
}

func downloadAgentBundleFromKubernetes(antreaClientset antrea.Interface, k8sClient kubernetes.Interface, failedNodes []string, dir string) error {
var errors []error
	agentInfoList, err := antreaClientset.CrdV1beta1().AntreaAgentInfos().List(context.TODO(), metav1.ListOptions{ResourceVersion: "0"})
	if err != nil {
		...
	}
        ...
}
```

Reply:
sounds better. will update this part.